lib/AST/ASTContext.cpp

```diff
@@ -2358,6 +2358,13 @@ Type ErrorType::get(Type originalType) {
   return entry = new (mem) ErrorType(ctx, originalType, properties);
 }
 
+Type HoleType::get(ASTContext &ctx, OriginatorType originator) {
+  assert(originator);
+  RecursiveTypeProperties properties = RecursiveTypeProperties::HasTypeHole;
+  auto arena = getArena(properties);
```

Comment:
I believe this is going to always allocate the type in the permanent arena, which is bad if it contains a pointer to a TypeVariableType. The recursive properties used to determine the arena should be the properties of the originator type, not the new properties of the hole type itself.

Reply:
I didn’t realize that, thank you, Slava!

Reply:
I need to make sure that types containing holes as well as hole types themselves are allocated in `ConstraintSolver` arena...

Reply:
Fixed! Every type which contains holes or is a `HoleType` itself is not going to be allocated in `ConstraintSolver` arena. @slavapestov WDYT?